tests/test_shapes.py

```diff
@@ -105,62 +106,89 @@ def test_shapes():
-            "offset": 0.3,
-            "cell_size": (1, 2, 1),
-            "repeat_cell": (1, 2, 1),
-            "resolution": 10,
-        },
-    )
-    _ = elem.surface
-    _ = elem.surface
-    _ = elem.skeletals
-    sheet = elem.sheet
-    _ = elem.sheet
-    assert 0 < sheet.extract_surface().volume < np.prod(elem.cell_size) * np.prod(elem.repeat_cell)
-    _ = elem.generateVtk(type_part="lower skeletal")
-    _ = elem.generateVtk(type_part="upper skeletal")
-    _ = elem.generateVtk(type_part="surface")
-    _ = elem.generate(type_part="sheet")
+@pytest.mark.parametrize("type_part", ["lower skeletal", "upper skeletal", "sheet"])
+def test_tpms_given_cadquery_vtk_shapes_volume_must_be_equivalent(type_part: str):
+    # Arrange
+    tpms = microgen.Tpms(
+        surface_function=microgen.surface_functions.gyroid,
+        offset=1.0,
+    )
 
-    def graded_density(x, y, z):
-        return x
-    elem = microgen.shape.tpms.CylindricalTpms(
-        radius=1.0,
-        surface_function=microgen.shape.surface_functions.schwarzD,
-        offset=graded_density,
-        phase_shift=(0.1, 0.2, 0.3),
-        cell_size=(1, 2, 1),
-        repeat_cell=(2, 1, 1),
-        resolution=20,
+    # Act
+    shape = tpms.generate(type_part=type_part, smoothing=0, verbose=False)
+    shape_check = tpms.generateVtk(type_part=type_part)
+
+    # Assert
+    assert np.isclose(shape.Volume(), np.abs(shape_check.volume), rtol=1e-2)
+
+
+@pytest.mark.parametrize("type_part", ["lower skeletal", "upper skeletal"])
+def test_tpms_given_cadquery_vtk_zero_offset_skeletals_volume_must_be_equivalent(type_part: str):
+    # Arrange
+    tpms = microgen.Tpms(
+        surface_function=microgen.surface_functions.schwarzP,
+        offset=0.0,
     )
-    # elem.generate(type_part="surface")
-    _ = elem.upper_skeletal
-    _ = elem.upper_skeletal
-    _ = elem.generateVtk(type_part="sheet")
 
-    elem = microgen.shape.tpms.CylindricalTpms(
-        radius=1.0,
-        surface_function=microgen.shape.surface_functions.schwarzD,
-        repeat_cell=(1, 10, 1),
+    # Act
+    shape = tpms.generate(type_part=type_part, smoothing=0, verbose=False)
+    shape_check = tpms.generateVtk(type_part=type_part)
+
+    # Assert
+    assert np.isclose(shape.Volume(), np.abs(shape_check.volume), rtol=1e-2)
+
+@pytest.mark.parametrize("surface", [func[0] for func in getmembers(microgen.surface_functions, isfunction)])
+def test_tpms_given_sum_volume_must_be_cube_volume(surface: str):
+    # Arrange
+    tpms = microgen.Tpms(
+        surface_function=getattr(microgen.surface_functions, surface),
+        offset=1.0,
+        repeat_cell=2,
+        cell_size=3.0,
     )
-    _ = elem.lower_skeletal
-    _ = elem.lower_skeletal
 
-    elem = microgen.shape.tpms.SphericalTpms(
+    # Act
+    volume = np.abs(tpms.sheet.volume + tpms.lower_skeletal.volume + tpms.upper_skeletal.volume)
+    cube_volume = np.prod(tpms.repeat_cell) * np.prod(tpms.cell_size)

```

Comment:
You can try to change `repeat_cell` and `cell_size` parameters values, using pytest.mark.parametrize (or using hypothesis strategies)  to ensure robustness.